ArgsParsing/TypeParsers/PkmnSpeciesParser.cs

```diff
@@ -1,38 +1,53 @@
 
 namespace ArgsParsing.TypeParsers
 {
     /// <summary>
     /// A parser that recognizes a pokemon species, either by name or by a #-prefixed species id.
-    /// Note that this class builds a name lookup at construction,
-    /// so make sure all pokemon species names are set up before instantiating this parser.
-    /// For details on that, see <see cref="PkmnSpecies.RegisterName"/>.
     /// </summary>
     public class PkmnSpeciesParser : BaseArgumentParser<PkmnSpecies>
     {
-        private readonly ImmutableSortedDictionary<string, PkmnSpecies> _lookup;
+        private readonly ImmutableDictionary<string, PkmnSpecies> _nameLookup;
+        private readonly IImmutableDictionary<string, PkmnSpecies> _idLookup;
 
-        public PkmnSpeciesParser()
+        /// <summary>
+        /// Create a new pkmn species parser for a set of known species.
+        /// </summary>
+        /// <param name="knownSpecies">all species that should be considered "existing" by the parser</param>
+        public PkmnSpeciesParser(IEnumerable<PkmnSpecies> knownSpecies)
         {
-            _lookup = PkmnSpecies.AllCurrentlyKnownSpecies.ToImmutableSortedDictionary(s => s.Name.ToLower(), s => s);
+            _nameLookup = knownSpecies.ToImmutableDictionary(s => NormalizeName(s.Name), s => s);
+            _idLookup = _nameLookup.Values.ToImmutableDictionary(s => s.Id, s => s);
         }
 
+        private static readonly Regex CharsToRemove = new Regex("[ '-.:]", RegexOptions.Compiled);
+        private static string NormalizeName(string name) => CharsToRemove.Replace(name.ToLowerInvariant(), "");
+
         public override Task<ArgsParseResult<PkmnSpecies>> Parse(IImmutableList<string> args, Type[] genericTypes)
         {
             if (!args[0].StartsWith("#"))
             {
-                // TODO Add some kind of name normalization to the lookup.
-                // TODO That's currently done by pokecat in the old core, so it will likely be some library.
-                string normalizedName = args[0].ToLower();
-                if (_lookup.TryGetValue(normalizedName, out PkmnSpecies speciesFromName))
+                string normalizedName = NormalizeName(args[0]);
+                if (_nameLookup.TryGetValue(normalizedName, out PkmnSpecies speciesFromName))
                 {
                     return Task.FromResult(ArgsParseResult<PkmnSpecies>.Success(
                         speciesFromName, args.Skip(1).ToImmutableList()));
                 }
+                if (args.Count >= 2)
+                {
+                    string normalizedNameTwoArgs = NormalizeName(args[0] + ' ' + args[1]);
+                    if (_nameLookup.TryGetValue(normalizedNameTwoArgs, out PkmnSpecies speciesFromTwoArgsName))
+                    {
+                        return Task.FromResult(ArgsParseResult<PkmnSpecies>.Success(
+                            speciesFromTwoArgsName, args.Skip(2).ToImmutableList()));
+                    }
+                }
```

Comment:
this new block has an interesting side-effect: I can put a space in-between a species name that does not actually have a space, e.g. `pid gey`, and it will be found. This feels very odd.
Off the top of my head I can think of 2 possible fixes:
- Use a less generic name normalization. This could be a hardcoded list of some name variations like `mr mime`, `mr-mime`, `mrmime` etc.
- add a special case here: if a pokemon was found while using 2 arguments, check if the found species' name actually contains a space. Otherwise treat it as a failure.

I think the former might be better, as in its current format there are countless other variations that really shouldn't match, e.g. `.:-'P'-:.:-'i'-:.:-'d'-:.:-'g'-:.:-'e'-:.:-'y'-:`

Reply:
the former would probably be preferable